// @remotion/bundler/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundle = exports.BundlerInternals = void 0;
const get_latest_remotion_version_1 = require("./get-latest-remotion-version");
const get_package_manager_1 = require("./get-package-manager");
const start_server_1 = require("./start-server");
const webpack_cache_1 = require("./webpack-cache");
const esbuild = require("esbuild");
const mime_types_1 = __importDefault(require("mime-types"));
exports.BundlerInternals = {
    startServer: start_server_1.startServer,
    cacheExists: webpack_cache_1.cacheExists,
    clearCache: webpack_cache_1.clearCache,
    getLatestRemotionVersion: get_latest_remotion_version_1.getLatestRemotionVersion,
    getPackageManager: get_package_manager_1.getPackageManager,
    lockFilePaths: get_package_manager_1.lockFilePaths,
    esbuild,
    mimeTypes: mime_types_1.default,
};
var bundler_1 = require("./bundler");
Object.defineProperty(exports, "bundle", { enumerable: true, get: function () { return bundler_1.bundle; } });
