// @remotion/bundler/dist/renderEntry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setBundleModeAndUpdate = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
// In React 18, you should use createRoot() from "react-dom/client".
// In React 18, you should use render from "react-dom".
// We support both, but Webpack chooses both of them and normalizes them to "react-dom/client",
// hence why we import the right thing all the time but need to differentiate here
const client_1 = __importDefault(require("react-dom/client"));
const remotion_1 = require("remotion");
const bundle_mode_1 = require("./bundle-mode");
const homepage_1 = require("./homepage/homepage");
remotion_1.Internals.CSSUtils.injectCSS(remotion_1.Internals.CSSUtils.makeDefaultCSS(null, '#fff'));
const Root = remotion_1.Internals.getRoot();
if (!Root) {
    throw new Error('Root has not been registered.');
}
const handle = (0, remotion_1.delayRender)('Loading root component');
const Fallback = () => {
    (0, react_1.useEffect)(() => {
        const fallback = (0, remotion_1.delayRender)('Waiting for Root component to unsuspend');
        return () => (0, remotion_1.continueRender)(fallback);
    }, []);
    return null;
};
const GetVideo = ({ state }) => {
    var _a;
    const video = remotion_1.Internals.useVideo();
    const compositions = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const [Component, setComponent] = (0, react_1.useState)(null);
    (0, react_1.useEffect)(() => {
        var _a;
        if (state.type !== 'composition') {
            return;
        }
        if (!video && compositions.compositions.length > 0) {
            const foundComposition = compositions.compositions.find((c) => c.id === state.compositionName);
            if (!foundComposition) {
                throw new Error('Found no composition with the name ' + state.compositionName);
            }
            compositions.setCurrentComposition((_a = foundComposition === null || foundComposition === void 0 ? void 0 : foundComposition.id) !== null && _a !== void 0 ? _a : null);
        }
    }, [compositions, compositions.compositions, state, video]);
    const fetchComponent = (0, react_1.useCallback)(() => {
        if (!video) {
            throw new Error('Expected to have video');
        }
        const Comp = video.component;
        setComponent(Comp);
    }, [video]);
    (0, react_1.useEffect)(() => {
        if (video) {
            fetchComponent();
        }
    }, [fetchComponent, video]);
    (0, react_1.useEffect)(() => {
        if (state.type === 'evaluation') {
            (0, remotion_1.continueRender)(handle);
        }
        else if (Component) {
            (0, remotion_1.continueRender)(handle);
        }
    }, [Component, state.type]);
    if (!video) {
        return null;
    }
    return ((0, jsx_runtime_1.jsx)(react_1.Suspense, { fallback: (0, jsx_runtime_1.jsx)(Fallback, {}, void 0), children: (0, jsx_runtime_1.jsx)("div", { id: "remotion-canvas", style: {
                width: video.width,
                height: video.height,
                display: 'flex',
                backgroundColor: 'transparent',
            }, children: Component ? ((0, jsx_runtime_1.jsx)(Component, { ...((_a = video === null || video === void 0 ? void 0 : video.defaultProps) !== null && _a !== void 0 ? _a : {}), ...(0, remotion_1.getInputProps)() }, void 0)) : null }, void 0) }, void 0));
};
const videoContainer = document.getElementById('video-container');
const explainerContainer = document.getElementById('explainer-container');
let cleanupVideoContainer = () => {
    videoContainer.innerHTML = '';
};
let cleanupExplainerContainer = () => {
    explainerContainer.innerHTML = '';
};
const renderContent = () => {
    const bundleMode = (0, bundle_mode_1.getBundleMode)();
    if (bundleMode.type === 'composition' || bundleMode.type === 'evaluation') {
        const markup = ((0, jsx_runtime_1.jsxs)(remotion_1.Internals.RemotionRoot, { children: [(0, jsx_runtime_1.jsx)(Root, {}, void 0), (0, jsx_runtime_1.jsx)(GetVideo, { state: bundleMode }, void 0)] }, void 0));
        if (client_1.default.createRoot) {
            const root = client_1.default.createRoot(videoContainer);
            root.render(markup);
            cleanupVideoContainer = () => {
                root.unmount();
            };
        }
        else {
            client_1.default.render(markup, videoContainer);
            cleanupVideoContainer = () => {
                client_1.default.unmountComponentAtNode(videoContainer);
            };
        }
    }
    else {
        cleanupVideoContainer();
        cleanupVideoContainer = () => {
            videoContainer.innerHTML = '';
        };
    }
    if (bundleMode.type === 'index' || bundleMode.type === 'evaluation') {
        if (client_1.default.createRoot) {
            const root = client_1.default.createRoot(explainerContainer);
            root.render((0, jsx_runtime_1.jsx)(homepage_1.Homepage, {}, void 0));
            cleanupExplainerContainer = () => {
                root.unmount();
            };
        }
        else {
            const root = client_1.default;
            root.render((0, jsx_runtime_1.jsx)(homepage_1.Homepage, {}, void 0), explainerContainer);
            cleanupExplainerContainer = () => {
                root.unmountComponentAtNode(explainerContainer);
            };
        }
    }
    else {
        cleanupExplainerContainer();
        cleanupExplainerContainer = () => {
            explainerContainer.innerHTML = '';
        };
    }
};
renderContent();
const setBundleModeAndUpdate = (state) => {
    (0, bundle_mode_1.setBundleMode)(state);
    renderContent();
};
exports.setBundleModeAndUpdate = setBundleModeAndUpdate;
if (typeof window !== 'undefined') {
    window.getStaticCompositions = () => {
        if (!remotion_1.Internals.compositionsRef.current) {
            throw new Error('Unexpectedly did not have a CompositionManager');
        }
        return remotion_1.Internals.compositionsRef.current
            .getCompositions()
            .map((c) => {
            return {
                defaultProps: c.defaultProps,
                durationInFrames: c.durationInFrames,
                fps: c.fps,
                height: c.height,
                id: c.id,
                width: c.width,
            };
        });
    };
    window.siteVersion = '2';
    window.setBundleMode = exports.setBundleModeAndUpdate;
}
