// @remotion/bundler/dist/update-available.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUpdateAvailableWithTimeout = void 0;
const semver_1 = __importDefault(require("semver"));
const get_latest_remotion_version_1 = require("./get-latest-remotion-version");
const get_package_manager_1 = require("./get-package-manager");
const isUpdateAvailable = async (currentVersion) => {
    const latest = await (0, get_latest_remotion_version_1.getLatestRemotionVersion)();
    return {
        updateAvailable: semver_1.default.lt(currentVersion, latest),
        currentVersion,
        latestVersion: latest,
        timedOut: false,
        packageManager: (0, get_package_manager_1.getPackageManager)(),
    };
};
const isUpdateAvailableWithTimeout = () => {
    const packageJson = require('../package.json');
    const { version } = packageJson;
    const threeSecTimeout = new Promise((resolve) => {
        setTimeout(() => {
            resolve({
                currentVersion: version,
                latestVersion: version,
                updateAvailable: false,
                timedOut: true,
                packageManager: (0, get_package_manager_1.getPackageManager)(),
            });
        }, 3000);
    });
    return Promise.race([threeSecTimeout, isUpdateAvailable(version)]);
};
exports.isUpdateAvailableWithTimeout = isUpdateAvailableWithTimeout;
