// @remotion/bundler/dist/dev-middleware/compatible-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.send = exports.setStatusCode = exports.setHeaderForResponse = exports.getHeaderFromResponse = exports.getHeaderFromRequest = exports.getHeaderNames = void 0;
function getHeaderNames(res) {
    if (typeof res.getHeaderNames !== 'function') {
        // @ts-expect-error
        return Object.keys(res._headers || {});
    }
    return res.getHeaderNames();
}
exports.getHeaderNames = getHeaderNames;
function getHeaderFromRequest(req, name) {
    // Express API
    if (typeof req.get === 'function') {
        return req.get(name);
    }
    return req.headers[name];
}
exports.getHeaderFromRequest = getHeaderFromRequest;
function getHeaderFromResponse(res, name) {
    return res.get(name);
}
exports.getHeaderFromResponse = getHeaderFromResponse;
function setHeaderForResponse(res, name, value) {
    res.set(name, typeof value === 'number' ? String(value) : value);
}
exports.setHeaderForResponse = setHeaderForResponse;
function setStatusCode(res, code) {
    res.status(code);
}
exports.setStatusCode = setStatusCode;
function send(req, res, bufferOtStream, byteLength) {
    if (typeof bufferOtStream === 'string' || Buffer.isBuffer(bufferOtStream)) {
        res.send(bufferOtStream);
        return;
    }
    setHeaderForResponse(res, 'Content-Length', byteLength);
    if (req.method === 'HEAD') {
        res.end();
        return;
    }
    bufferOtStream.pipe(res);
}
exports.send = send;
