// @remotion/cli/dist/editor/components/CompositionSelectorItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompositionSelectorItem = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../helpers/colors");
const is_composition_still_1 = require("../helpers/is-composition-still");
const film_1 = require("../icons/film");
const still_1 = require("../icons/still");
const layout_1 = require("./layout");
const item = {
    paddingLeft: 8,
    paddingRight: 8,
    paddingTop: 6,
    paddingBottom: 6,
    fontSize: 13,
    display: 'flex',
    textDecoration: 'none',
    cursor: 'default',
    alignItems: 'center',
    marginBottom: 1,
};
const iconStyle = {
    width: 18,
    height: 18,
};
const CompositionSelectorItem = ({ composition, currentComposition, tabIndex, selectComposition }) => {
    const selected = currentComposition === composition.id;
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const onPointerEnter = (0, react_1.useCallback)(() => {
        setHovered(true);
    }, []);
    const onPointerLeave = (0, react_1.useCallback)(() => {
        setHovered(false);
    }, []);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...item,
            backgroundColor: hovered
                ? selected
                    ? colors_1.SELECTED_BACKGROUND
                    : colors_1.CLEAR_HOVER
                : selected
                    ? colors_1.SELECTED_BACKGROUND
                    : 'transparent',
            color: selected || hovered ? 'white' : colors_1.LIGHT_TEXT,
        };
    }, [hovered, selected]);
    const onClick = (0, react_1.useCallback)((evt) => {
        evt.preventDefault();
        selectComposition(composition);
    }, [composition, selectComposition]);
    return ((0, jsx_runtime_1.jsxs)("a", { style: style, onPointerEnter: onPointerEnter, onPointerLeave: onPointerLeave, href: composition.id, tabIndex: tabIndex, onClick: onClick, children: [(0, is_composition_still_1.isCompositionStill)(composition) ? ((0, jsx_runtime_1.jsx)(still_1.StillIcon, { style: iconStyle }, void 0)) : ((0, jsx_runtime_1.jsx)(film_1.FilmIcon, { style: iconStyle }, void 0)), (0, jsx_runtime_1.jsx)(layout_1.Spacing, { x: 1 }, void 0), composition.id] }, void 0));
};
exports.CompositionSelectorItem = CompositionSelectorItem;
