// @remotion/cli/dist/editor/components/CompositionSelector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompositionSelector = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const marks_1 = require("../state/marks");
const z_index_1 = require("../state/z-index");
const CompositionSelectorItem_1 = require("./CompositionSelectorItem");
const CurrentComposition_1 = require("./CurrentComposition");
const FramePersistor_1 = require("./FramePersistor");
const TimelineInOutToggle_1 = require("./TimelineInOutToggle");
const container = {
    borderRight: '1px solid black',
    position: 'absolute',
    height: '100%',
    width: '100%',
    flex: 1,
};
const list = {
    padding: 5,
    height: 'calc(100% - 100px)',
    overflowY: 'auto',
};
const CompositionSelector = () => {
    const { compositions, setCurrentComposition, currentComposition } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const { tabIndex } = (0, z_index_1.useZIndex)();
    const setCurrentFrame = remotion_1.Internals.Timeline.useTimelineSetFrame();
    const selectComposition = (0, react_1.useCallback)((c) => {
        var _a;
        (_a = TimelineInOutToggle_1.inOutHandles.current) === null || _a === void 0 ? void 0 : _a.setMarks((0, marks_1.loadMarks)(c.id, c.durationInFrames));
        window.history.pushState({}, 'Preview', `/${c.id}`);
        const frame = (0, FramePersistor_1.getFrameForComposition)(c.id);
        const frameInBounds = Math.min(c.durationInFrames - 1, frame);
        setCurrentFrame(frameInBounds);
        setCurrentComposition(c.id);
    }, [setCurrentComposition, setCurrentFrame]);
    (0, react_1.useEffect)(() => {
        if (currentComposition) {
            return;
        }
        const compositionFromUrl = (0, FramePersistor_1.getCurrentCompositionFromUrl)();
        if (compositionFromUrl) {
            const exists = compositions.find((c) => c.id === compositionFromUrl);
            if (exists) {
                selectComposition(exists);
                return;
            }
        }
        if (compositions.length > 0) {
            selectComposition(compositions[0]);
        }
    }, [compositions, currentComposition, selectComposition]);
    return ((0, jsx_runtime_1.jsxs)("div", { style: container, children: [(0, jsx_runtime_1.jsx)(CurrentComposition_1.CurrentComposition, {}, void 0), (0, jsx_runtime_1.jsx)("div", { style: list, children: compositions.map((c) => {
                    return ((0, jsx_runtime_1.jsx)(CompositionSelectorItem_1.CompositionSelectorItem, { currentComposition: currentComposition, selectComposition: selectComposition, tabIndex: tabIndex, composition: c }, c.id));
                }) }, void 0)] }, void 0));
};
exports.CompositionSelector = CompositionSelector;
