// @remotion/cli/dist/editor/components/MediaVolumeSlider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaVolumeSlider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const player_1 = require("@remotion/player");
const react_1 = require("react");
const remotion_1 = require("remotion");
const media_volume_1 = require("../icons/media-volume");
const ControlButton_1 = require("./ControlButton");
const BAR_HEIGHT = 5;
const KNOB_SIZE = 12;
const VERTICAL_PADDING = 4;
const containerStyle = {
    userSelect: 'none',
    paddingTop: VERTICAL_PADDING,
    paddingBottom: VERTICAL_PADDING,
    boxSizing: 'border-box',
    cursor: 'pointer',
    position: 'relative',
};
const barBackground = {
    height: BAR_HEIGHT,
    backgroundColor: 'rgba(255, 255, 255, 0.5)',
    width: 100,
    borderRadius: BAR_HEIGHT / 2,
};
const getVolumeFromX = (clientX, width) => {
    const pos = clientX;
    const volume = remotion_1.interpolate(pos, [0, width], [0, 1], {
        extrapolateLeft: 'clamp',
        extrapolateRight: 'clamp',
    });
    return volume;
};
const xSpacer = {
    width: 5,
};
const MediaVolumeSlider = () => {
    const [mediaMuted, setMediaMuted] = remotion_1.Internals.useMediaMutedState();
    const [mediaVolume, setMediaVolume] = remotion_1.Internals.useMediaVolumeState();
    const [dragging, setDragging] = react_1.useState(false);
    const currentRef = react_1.useRef(null);
    const iconDivRef = react_1.useRef(null);
    const parentDivRef = react_1.useRef(null);
    const size = player_1.PlayerInternals.useElementSize(currentRef);
    const hover = player_1.PlayerInternals.useHoverState(parentDivRef);
    const onClick = react_1.useCallback(() => {
        setMediaMuted((muted) => !muted);
        setMediaVolume(Number(mediaMuted));
    }, [setMediaMuted, mediaMuted, setMediaVolume]);
    const onPointerDown = react_1.useCallback((e) => {
        if (!size) {
            throw new Error('Player has no size');
        }
        const _volume = getVolumeFromX(e.clientX - size.left, size.width);
        setMediaVolume(_volume);
        setMediaMuted(_volume <= 0);
        setDragging(true);
    }, [setMediaMuted, setMediaVolume, size]);
    const onPointerMove = react_1.useCallback((e) => {
        if (!size) {
            throw new Error('Player has no size');
        }
        if (!dragging)
            return;
        const _volume = getVolumeFromX(e.clientX - size.left, size.width);
        setMediaVolume(_volume);
        setMediaMuted(_volume <= 0);
    }, [dragging, setMediaMuted, setMediaVolume, size]);
    const onPointerUp = react_1.useCallback(() => {
        console.log('stop dragging');
        setDragging(false);
    }, []);
    react_1.useEffect(() => {
        if (!dragging) {
            return;
        }
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp);
        return () => {
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
        };
    }, [dragging, onPointerMove, onPointerUp]);
    const knobStyle = react_1.useMemo(() => {
        var _a;
        return {
            height: KNOB_SIZE,
            width: KNOB_SIZE,
            borderRadius: KNOB_SIZE / 2,
            position: 'absolute',
            top: VERTICAL_PADDING - KNOB_SIZE / 2 + 5 / 2,
            backgroundColor: 'white',
            left: Math.max(-KNOB_SIZE / 2, mediaVolume * ((_a = size === null || size === void 0 ? void 0 : size.width) !== null && _a !== void 0 ? _a : 0) - KNOB_SIZE / 2),
            boxShadow: '0 0 2px black',
            opacity: Number(hover),
        };
    }, [hover, mediaVolume, size === null || size === void 0 ? void 0 : size.width]);
    const fillStyle = react_1.useMemo(() => {
        return {
            height: BAR_HEIGHT,
            backgroundColor: 'rgba(255, 255, 255, 1)',
            width: (mediaVolume / 1) * 100 + '%',
            borderRadius: BAR_HEIGHT / 2,
        };
    }, [mediaVolume]);
    return (jsx_runtime_1.jsxs(ControlButton_1.ControlDiv, Object.assign({ ref: parentDivRef }, { children: [jsx_runtime_1.jsx("div", Object.assign({ ref: iconDivRef, onClick: onClick }, { children: mediaMuted ? jsx_runtime_1.jsx(media_volume_1.VolumeOffIcon, {}, void 0) : jsx_runtime_1.jsx(media_volume_1.VolumeOnIcon, {}, void 0) }), void 0),
            jsx_runtime_1.jsx("div", { style: xSpacer }, void 0),
            jsx_runtime_1.jsx("div", Object.assign({ ref: currentRef, onPointerDown: onPointerDown, style: containerStyle }, { children: hover ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [jsx_runtime_1.jsx("div", Object.assign({ style: barBackground }, { children: jsx_runtime_1.jsx("div", { style: fillStyle }, void 0) }), void 0),
                        jsx_runtime_1.jsx("div", { style: knobStyle }, void 0)] }, void 0)) : (jsx_runtime_1.jsx("div", { style: { width: 100 } }, void 0)) }), void 0)] }), void 0));
};
exports.MediaVolumeSlider = MediaVolumeSlider;
