// @remotion/cli/dist/editor/components/NewSequence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewSequence = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../helpers/colors");
const font_1 = require("../helpers/font");
const backgroundOverlay = {
    backgroundColor: 'rgba(255, 255, 255, 0.05)',
    position: 'fixed',
    height: '100%',
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
};
const panel = {
    backgroundColor: colors_1.BACKGROUND,
    boxShadow: '0 0 4px black',
    padding: 20,
    color: 'white',
    fontFamily: font_1.FONT_FAMILY,
};
const NewSequence = () => {
    const [type, setType] = (0, react_1.useState)('composition');
    const [width, setWidth] = (0, react_1.useState)('1280');
    const [height, setHeight] = (0, react_1.useState)('720');
    const onStillChanged = (0, react_1.useCallback)((e) => {
        if (e.target.checked) {
            setType('still');
        }
    }, []);
    const onCompositionChanged = (0, react_1.useCallback)((e) => {
        if (e.target.checked) {
            setType('composition');
        }
    }, []);
    const onWidthChanged = (0, react_1.useCallback)((e) => {
        setWidth(e.target.value);
    }, []);
    const onHeightChanged = (0, react_1.useCallback)((e) => {
        setHeight(e.target.value);
    }, []);
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({ style: backgroundOverlay }, { children: (0, jsx_runtime_1.jsx)("div", Object.assign({ style: panel }, { children: (0, jsx_runtime_1.jsxs)("form", { children: [(0, jsx_runtime_1.jsx)("input", { type: "text", placeholder: "Composition name" }, void 0), (0, jsx_runtime_1.jsx)("br", {}, void 0), (0, jsx_runtime_1.jsx)("br", {}, void 0), (0, jsx_runtime_1.jsxs)("label", { children: [(0, jsx_runtime_1.jsx)("input", { onChange: onCompositionChanged, checked: type === 'composition', name: "type", type: "radio" }, void 0), "Composition"] }, void 0), (0, jsx_runtime_1.jsxs)("label", { children: [(0, jsx_runtime_1.jsx)("input", { onChange: onStillChanged, checked: type === 'still', name: "type", type: "radio" }, void 0), "Still"] }, void 0), (0, jsx_runtime_1.jsx)("br", {}, void 0), (0, jsx_runtime_1.jsx)("br", {}, void 0), (0, jsx_runtime_1.jsx)("input", { type: "number", value: width, placeholder: "Width (px)", onChange: onWidthChanged, name: "width" }, void 0), (0, jsx_runtime_1.jsx)("br", {}, void 0), (0, jsx_runtime_1.jsx)("input", { type: "number", value: height, onChange: onHeightChanged, placeholder: "Height (px)", name: "height" }, void 0), (0, jsx_runtime_1.jsx)("br", {}, void 0), (0, jsx_runtime_1.jsxs)("select", { children: [(0, jsx_runtime_1.jsx)("option", { children: "24fps" }, void 0), (0, jsx_runtime_1.jsx)("option", { children: "25fps" }, void 0), (0, jsx_runtime_1.jsx)("option", { children: "29.97fps" }, void 0), (0, jsx_runtime_1.jsx)("option", { children: "30fps" }, void 0), (0, jsx_runtime_1.jsx)("option", { children: "48fps" }, void 0), (0, jsx_runtime_1.jsx)("option", { children: "50fps" }, void 0), (0, jsx_runtime_1.jsx)("option", { children: "60fps" }, void 0)] }, void 0)] }, void 0) }), void 0) }), void 0));
};
exports.NewSequence = NewSequence;
