// @remotion/renderer/dist/get-compositions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCompositions = void 0;
const get_browser_instance_1 = require("./get-browser-instance");
const handle_javascript_exception_1 = require("./error-handling/handle-javascript-exception");
const prepare_server_1 = require("./prepare-server");
const puppeteer_evaluate_1 = require("./puppeteer-evaluate");
const set_props_and_env_1 = require("./set-props-and-env");
const validate_puppeteer_timeout_1 = require("./validate-puppeteer-timeout");
const innerGetCompositions = async (serveUrl, page, config) => {
    if (config === null || config === void 0 ? void 0 : config.onBrowserLog) {
        page.on('console', (log) => {
            var _a;
            (_a = config.onBrowserLog) === null || _a === void 0 ? void 0 : _a.call(config, {
                stackTrace: log.stackTrace(),
                text: log.text(),
                type: log.type(),
            });
        });
    }
    (0, validate_puppeteer_timeout_1.validatePuppeteerTimeout)(config === null || config === void 0 ? void 0 : config.timeoutInMilliseconds);
    await (0, set_props_and_env_1.setPropsAndEnv)({
        inputProps: config === null || config === void 0 ? void 0 : config.inputProps,
        envVariables: config === null || config === void 0 ? void 0 : config.envVariables,
        page,
        serveUrl,
        initialFrame: 0,
        timeoutInMilliseconds: config === null || config === void 0 ? void 0 : config.timeoutInMilliseconds,
    });
    await (0, puppeteer_evaluate_1.puppeteerEvaluateWithCatch)({
        page,
        pageFunction: () => {
            window.setBundleMode({
                type: 'evaluation',
            });
        },
        frame: null,
        args: [],
    });
    await page.waitForFunction('window.ready === true');
    const result = await (0, puppeteer_evaluate_1.puppeteerEvaluateWithCatch)({
        pageFunction: () => {
            return window.getStaticCompositions();
        },
        frame: null,
        page,
        args: [],
    });
    return result;
};
const getCompositions = async (serveUrlOrWebpackUrl, config) => {
    var _a, _b;
    const { serveUrl, closeServer } = await (0, prepare_server_1.prepareServer)(serveUrlOrWebpackUrl);
    const { page, cleanup } = await (0, get_browser_instance_1.getPageAndCleanupFn)({
        passedInInstance: config === null || config === void 0 ? void 0 : config.puppeteerInstance,
        browserExecutable: (_a = config === null || config === void 0 ? void 0 : config.browserExecutable) !== null && _a !== void 0 ? _a : null,
        chromiumOptions: (_b = config === null || config === void 0 ? void 0 : config.chromiumOptions) !== null && _b !== void 0 ? _b : {},
    });
    return new Promise((resolve, reject) => {
        const cleanupPageError = (0, handle_javascript_exception_1.handleJavascriptException)({
            page,
            frame: null,
            onError: (err) => reject(err),
        });
        innerGetCompositions(serveUrl, page, config !== null && config !== void 0 ? config : {})
            .then((comp) => resolve(comp))
            .catch((err) => {
            reject(err);
        })
            .finally(() => {
            cleanup();
            closeServer();
            cleanupPageError();
        });
    });
};
exports.getCompositions = getCompositions;
