// @remotion/renderer/dist/handle-javascript-exception.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleJavascriptException = exports.removeDelayRenderStack = exports.ErrorWithStackFrame = void 0;
const remotion_1 = require("remotion");
const delay_render_embedded_stack_1 = require("./delay-render-embedded-stack");
const symbolicate_stacktrace_1 = require("./symbolicate-stacktrace");
class ErrorWithStackFrame extends Error {
    constructor({ message, _stackFrames, _frame, _errorType, _delayRenderCall, }) {
        super(message);
        this.stackFrames = _stackFrames;
        this.frame = _frame;
        this.errorType = _errorType;
        this.delayRenderCall = _delayRenderCall;
    }
}
exports.ErrorWithStackFrame = ErrorWithStackFrame;
const cleanUpErrorMessage = (exception) => {
    var _a, _b, _c, _d;
    let errorMessage = (_a = exception.exceptionDetails.exception) === null || _a === void 0 ? void 0 : _a.description;
    const errorType = (_b = exception.exceptionDetails.exception) === null || _b === void 0 ? void 0 : _b.className;
    const prefix = `${errorType}: `;
    if (errorMessage.startsWith(prefix)) {
        errorMessage = errorMessage.substring(prefix.length);
    }
    const frames = (_d = (_c = exception.exceptionDetails.stackTrace) === null || _c === void 0 ? void 0 : _c.callFrames.length) !== null && _d !== void 0 ? _d : 0;
    const split = errorMessage.split('\n');
    return split.slice(0, split.length - frames).join('\n');
};
const removeDelayRenderStack = (message) => {
    const index = message.indexOf(remotion_1.DELAY_RENDER_CALLSTACK_TOKEN);
    if (index === -1) {
        return message;
    }
    return message.substring(0, index);
};
exports.removeDelayRenderStack = removeDelayRenderStack;
const callFrameToStackFrame = (callFrame) => {
    return {
        columnNumber: callFrame.columnNumber,
        fileName: callFrame.url,
        functionName: callFrame.functionName,
        lineNumber: callFrame.lineNumber,
    };
};
const handleJavascriptException = ({ page, onError, frame, }) => {
    const client = page._client;
    const handler = async (exception) => {
        var _a, _b;
        const rawErrorMessage = (_a = exception.exceptionDetails.exception) === null || _a === void 0 ? void 0 : _a.description;
        const cleanErrorMessage = cleanUpErrorMessage(exception);
        const err = new Error((0, exports.removeDelayRenderStack)(cleanErrorMessage));
        err.stack = rawErrorMessage;
        if (!exception.exceptionDetails.stackTrace) {
            onError(err);
            return;
        }
        const errorType = (_b = exception.exceptionDetails.exception) === null || _b === void 0 ? void 0 : _b.className;
        const delayRenderStack = await (0, delay_render_embedded_stack_1.parseDelayRenderEmbeddedStack)(cleanErrorMessage);
        try {
            const sym = await (0, symbolicate_stacktrace_1.symbolicateStackTrace)(exception.exceptionDetails.stackTrace
                .callFrames.map((f) => callFrameToStackFrame(f)));
            const symbolicatedErr = new ErrorWithStackFrame({
                message: (0, exports.removeDelayRenderStack)(cleanErrorMessage),
                _stackFrames: sym,
                _frame: frame,
                _errorType: errorType,
                _delayRenderCall: delayRenderStack,
            });
            onError(symbolicatedErr);
        }
        catch (errorVisualizingError) {
            console.log('error symbolicating error:', errorVisualizingError);
            onError(err);
        }
    };
    client.on('Runtime.exceptionThrown', handler);
    return () => {
        client.off('Runtime.exceptionThrown', handler);
    };
};
exports.handleJavascriptException = handleJavascriptException;
