// @remotion/renderer/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderInternals = exports.openBrowser = exports.stitchFramesToVideo = exports.renderStill = exports.renderFrames = exports.getCompositions = exports.combineVideos = exports.renderMedia = exports.ErrorWithStackFrame = void 0;
const add_silent_audio_if_necessary_1 = require("./add-silent-audio-if-necessary");
const ffmpeg_flags_1 = require("./ffmpeg-flags");
const get_concurrency_1 = require("./get-concurrency");
const get_extension_from_codec_1 = require("./get-extension-from-codec");
const get_local_browser_executable_1 = require("./get-local-browser-executable");
const make_assets_download_dir_1 = require("./make-assets-download-dir");
const normalize_serve_url_1 = require("./normalize-serve-url");
const serve_static_1 = require("./serve-static");
const stitcher_1 = require("./stitcher");
const validate_even_dimensions_with_codec_1 = require("./validate-even-dimensions-with-codec");
const validate_ffmpeg_1 = require("./validate-ffmpeg");
const tmp_dir_1 = require("./tmp-dir");
const delete_directory_1 = require("./delete-directory");
const prepare_server_1 = require("./prepare-server");
const is_serve_url_1 = require("./is-serve-url");
const ensure_output_directory_1 = require("./ensure-output-directory");
const validate_puppeteer_timeout_1 = require("./validate-puppeteer-timeout");
const get_frame_to_render_1 = require("./get-frame-to-render");
const download_file_1 = require("./assets/download-file");
const validate_scale_1 = require("./validate-scale");
const open_browser_1 = require("./open-browser");
const parse_browser_error_stack_1 = require("./parse-browser-error-stack");
const symbolicate_error_1 = require("./error-handling/symbolicate-error");
const symbolicateable_error_1 = require("./error-handling/symbolicateable-error");
var handle_javascript_exception_1 = require("./error-handling/handle-javascript-exception");
Object.defineProperty(exports, "ErrorWithStackFrame", { enumerable: true, get: function () { return handle_javascript_exception_1.ErrorWithStackFrame; } });
var render_media_1 = require("./render-media");
Object.defineProperty(exports, "renderMedia", { enumerable: true, get: function () { return render_media_1.renderMedia; } });
var combine_videos_1 = require("./combine-videos");
Object.defineProperty(exports, "combineVideos", { enumerable: true, get: function () { return combine_videos_1.combineVideos; } });
var get_compositions_1 = require("./get-compositions");
Object.defineProperty(exports, "getCompositions", { enumerable: true, get: function () { return get_compositions_1.getCompositions; } });
var render_1 = require("./render");
Object.defineProperty(exports, "renderFrames", { enumerable: true, get: function () { return render_1.renderFrames; } });
var render_still_1 = require("./render-still");
Object.defineProperty(exports, "renderStill", { enumerable: true, get: function () { return render_still_1.renderStill; } });
var stitcher_2 = require("./stitcher");
Object.defineProperty(exports, "stitchFramesToVideo", { enumerable: true, get: function () { return stitcher_2.stitchFramesToVideo; } });
var open_browser_2 = require("./open-browser");
Object.defineProperty(exports, "openBrowser", { enumerable: true, get: function () { return open_browser_2.openBrowser; } });
exports.RenderInternals = {
    ensureLocalBrowser: get_local_browser_executable_1.ensureLocalBrowser,
    ffmpegHasFeature: ffmpeg_flags_1.ffmpegHasFeature,
    getActualConcurrency: get_concurrency_1.getActualConcurrency,
    getFfmpegVersion: ffmpeg_flags_1.getFfmpegVersion,
    validateFfmpeg: validate_ffmpeg_1.validateFfmpeg,
    binaryExists: validate_ffmpeg_1.binaryExists,
    getFfmpegBuildInfo: ffmpeg_flags_1.getFfmpegBuildInfo,
    serveStatic: serve_static_1.serveStatic,
    addSilentAudioIfNecessary: add_silent_audio_if_necessary_1.addSilentAudioIfNecessary,
    validateEvenDimensionsWithCodec: validate_even_dimensions_with_codec_1.validateEvenDimensionsWithCodec,
    normalizeServeUrl: normalize_serve_url_1.normalizeServeUrl,
    spawnFfmpeg: stitcher_1.spawnFfmpeg,
    getFileExtensionFromCodec: get_extension_from_codec_1.getFileExtensionFromCodec,
    makeAssetsDownloadTmpDir: make_assets_download_dir_1.makeAssetsDownloadTmpDir,
    tmpDir: tmp_dir_1.tmpDir,
    deleteDirectory: delete_directory_1.deleteDirectory,
    prepareServer: prepare_server_1.prepareServer,
    isServeUrl: is_serve_url_1.isServeUrl,
    ensureOutputDirectory: ensure_output_directory_1.ensureOutputDirectory,
    getRealFrameRange: get_frame_to_render_1.getRealFrameRange,
    validatePuppeteerTimeout: validate_puppeteer_timeout_1.validatePuppeteerTimeout,
    downloadFile: download_file_1.downloadFile,
    validateScale: validate_scale_1.validateScale,
    killAllBrowsers: open_browser_1.killAllBrowsers,
    parseStack: parse_browser_error_stack_1.parseStack,
    symbolicateError: symbolicate_error_1.symbolicateError,
    SymbolicateableError: symbolicateable_error_1.SymbolicateableError,
};
