// @remotion/renderer/dist/render.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderFrames = exports.innerRenderFrames = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const remotion_1 = require("remotion");
const download_and_map_assets_to_file_1 = require("./assets/download-and-map-assets-to-file");
const cycle_browser_tabs_1 = require("./cycle-browser-tabs");
const get_concurrency_1 = require("./get-concurrency");
const get_frame_to_render_1 = require("./get-frame-to-render");
const handle_javascript_exception_1 = require("./error-handling/handle-javascript-exception");
const image_format_1 = require("./image-format");
const legacy_webpack_config_1 = require("./legacy-webpack-config");
const make_assets_download_dir_1 = require("./make-assets-download-dir");
const open_browser_1 = require("./open-browser");
const pool_1 = require("./pool");
const prepare_server_1 = require("./prepare-server");
const provide_screenshot_1 = require("./provide-screenshot");
const seek_to_frame_1 = require("./seek-to-frame");
const set_props_and_env_1 = require("./set-props-and-env");
const validate_scale_1 = require("./validate-scale");
const puppeteer_evaluate_1 = require("./puppeteer-evaluate");
const getComposition = (others) => {
    if ('composition' in others) {
        return others.composition;
    }
    if ('config' in others) {
        return others.config;
    }
    return undefined;
};
const innerRenderFrames = async ({ parallelism, onFrameUpdate, outputDir, onStart, inputProps, quality, imageFormat = image_format_1.DEFAULT_IMAGE_FORMAT, frameRange, puppeteerInstance, onError, envVariables, onBrowserLog, onFrameBuffer, onDownload, pagesArray, serveUrl, composition, timeoutInMilliseconds, scale, }) => {
    if (!puppeteerInstance) {
        throw new Error('weird');
    }
    const actualParallelism = (0, get_concurrency_1.getActualConcurrency)(parallelism !== null && parallelism !== void 0 ? parallelism : null);
    if (outputDir) {
        if (!fs_1.default.existsSync(outputDir)) {
            fs_1.default.mkdirSync(outputDir, {
                recursive: true,
            });
        }
    }
    const realFrameRange = (0, get_frame_to_render_1.getRealFrameRange)(composition.durationInFrames, frameRange !== null && frameRange !== void 0 ? frameRange : null);
    const frameCount = realFrameRange[1] - realFrameRange[0] + 1;
    const pages = new Array(actualParallelism).fill(true).map(async () => {
        const page = await puppeteerInstance.newPage();
        pagesArray.push(page);
        page.setViewport({
            width: composition.width,
            height: composition.height,
            deviceScaleFactor: scale !== null && scale !== void 0 ? scale : 1,
        });
        const logCallback = (log) => {
            onBrowserLog === null || onBrowserLog === void 0 ? void 0 : onBrowserLog({
                stackTrace: log.stackTrace(),
                text: log.text(),
                type: log.type(),
            });
        };
        if (onBrowserLog) {
            page.on('console', logCallback);
        }
        const initialFrame = typeof frameRange === 'number'
            ? frameRange
            : frameRange === null || frameRange === undefined
                ? 0
                : frameRange[0];
        await (0, set_props_and_env_1.setPropsAndEnv)({
            inputProps,
            envVariables,
            page,
            serveUrl,
            initialFrame,
            timeoutInMilliseconds,
        });
        await (0, puppeteer_evaluate_1.puppeteerEvaluateWithCatch)({
            pageFunction: (id) => {
                window.setBundleMode({
                    type: 'composition',
                    compositionName: id,
                });
            },
            args: [composition.id],
            frame: null,
            page,
        });
        page.off('console', logCallback);
        return page;
    });
    const puppeteerPages = await Promise.all(pages);
    const pool = new pool_1.Pool(puppeteerPages);
    const [firstFrameIndex, lastFrameIndex] = realFrameRange;
    // Substract one because 100 frames will be 00-99
    // --> 2 digits
    const filePadLength = String(lastFrameIndex).length;
    let framesRendered = 0;
    onStart({
        frameCount,
    });
    const downloadDir = (0, make_assets_download_dir_1.makeAssetsDownloadTmpDir)();
    const assets = new Array(frameCount).fill(undefined);
    await Promise.all(new Array(frameCount)
        .fill(Boolean)
        .map((x, i) => i)
        .map(async (index) => {
        const frame = realFrameRange[0] + index;
        const freePage = await pool.acquire();
        const paddedIndex = String(frame).padStart(filePadLength, '0');
        const errorCallbackOnFrame = (err) => {
            onError(err);
        };
        const cleanupPageError = (0, handle_javascript_exception_1.handleJavascriptException)({
            page: freePage,
            onError: errorCallbackOnFrame,
            frame,
        });
        freePage.on('error', errorCallbackOnFrame);
        await (0, seek_to_frame_1.seekToFrame)({ frame, page: freePage });
        if (imageFormat !== 'none') {
            if (onFrameBuffer) {
                const buffer = await (0, provide_screenshot_1.provideScreenshot)({
                    page: freePage,
                    imageFormat,
                    quality,
                    options: {
                        frame,
                        output: undefined,
                    },
                });
                onFrameBuffer(buffer, frame);
            }
            else {
                if (!outputDir) {
                    throw new Error('Called renderFrames() without specifying either `outputDir` or `onFrameBuffer`');
                }
                const output = path_1.default.join(outputDir, `element-${paddedIndex}.${imageFormat}`);
                await (0, provide_screenshot_1.provideScreenshot)({
                    page: freePage,
                    imageFormat,
                    quality,
                    options: {
                        frame,
                        output,
                    },
                });
            }
        }
        const collectedAssets = await (0, puppeteer_evaluate_1.puppeteerEvaluateWithCatch)({
            pageFunction: () => {
                return window.remotion_collectAssets();
            },
            args: [],
            frame,
            page: freePage,
        });
        const compressedAssets = collectedAssets.map((asset) => remotion_1.Internals.AssetCompression.compressAsset(assets.filter(remotion_1.Internals.truthy).flat(1), asset));
        assets[index] = compressedAssets;
        compressedAssets.forEach((asset) => {
            (0, download_and_map_assets_to_file_1.downloadAndMapAssetsToFileUrl)({
                asset,
                downloadDir,
                onDownload: onDownload !== null && onDownload !== void 0 ? onDownload : (() => () => undefined),
            }).catch((err) => {
                onError(new Error(`Error while downloading asset: ${err.stack}`));
            });
        });
        pool.release(freePage);
        framesRendered++;
        onFrameUpdate(framesRendered, frame);
        cleanupPageError();
        freePage.off('error', errorCallbackOnFrame);
        return compressedAssets;
    }));
    const returnValue = {
        assetsInfo: {
            assets,
            downloadDir,
            firstFrameIndex,
            imageSequenceName: `element-%0${filePadLength}d.${imageFormat}`,
        },
        frameCount,
    };
    return returnValue;
};
exports.innerRenderFrames = innerRenderFrames;
const renderFrames = async (options) => {
    var _a;
    const composition = getComposition(options);
    if (!composition) {
        throw new Error('No `composition` option has been specified for renderFrames()');
    }
    remotion_1.Internals.validateDimension(composition.height, 'height', 'in the `config` object passed to `renderFrames()`');
    remotion_1.Internals.validateDimension(composition.width, 'width', 'in the `config` object passed to `renderFrames()`');
    remotion_1.Internals.validateFps(composition.fps, 'in the `config` object of `renderFrames()`');
    remotion_1.Internals.validateDurationInFrames(composition.durationInFrames, 'in the `config` object passed to `renderFrames()`');
    if (options.quality !== undefined && options.imageFormat !== 'jpeg') {
        throw new Error("You can only pass the `quality` option if `imageFormat` is 'jpeg'.");
    }
    const selectedServeUrl = (0, legacy_webpack_config_1.getServeUrlWithFallback)(options);
    remotion_1.Internals.validateQuality(options.quality);
    (0, validate_scale_1.validateScale)(options.scale);
    const { closeServer, serveUrl } = await (0, prepare_server_1.prepareServer)(selectedServeUrl);
    const browserInstance = (_a = options.puppeteerInstance) !== null && _a !== void 0 ? _a : (await (0, open_browser_1.openBrowser)(remotion_1.Internals.DEFAULT_BROWSER, {
        shouldDumpIo: options.dumpBrowserLogs,
        browserExecutable: options.browserExecutable,
        chromiumOptions: options.chromiumOptions,
    }));
    const { stopCycling } = (0, cycle_browser_tabs_1.cycleBrowserTabs)(browserInstance);
    const openedPages = [];
    return new Promise((resolve, reject) => {
        (0, exports.innerRenderFrames)({
            ...options,
            puppeteerInstance: browserInstance,
            onError: (err) => reject(err),
            pagesArray: openedPages,
            serveUrl,
            composition,
        })
            .then((res) => resolve(res))
            .catch((err) => reject(err))
            .finally(() => {
            // If browser instance was passed in, we close all the pages
            // we opened.
            // If new browser was opened, then closing the browser as a cleanup.
            if (options.puppeteerInstance) {
                Promise.all(openedPages.map((p) => p.close())).catch((err) => {
                    console.log('Unable to close browser tab', err);
                });
            }
            else {
                browserInstance.close().catch((err) => {
                    console.log('Unable to close browser', err);
                });
            }
            stopCycling();
            closeServer();
        });
    });
};
exports.renderFrames = renderFrames;
