// @remotion/renderer/dist/render-video.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderVideo = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const remotion_1 = require("remotion");
const stitcher_1 = require("./stitcher");
const render_1 = require("./render");
const renderVideo = async ({ parallelism, proResProfile, parallelEncoding, crf, outputDir, config, imageFormat, ffmpegExecutable, inputProps, pixelFormat, codec, envVariables, quality, frameRange, browser, serveUrl, openedBrowser, absoluteOutputFile, onProgress, overwrite, shouldOutputImageSequence, fileExtension, bundled, onDownload, }) => {
    var _a, _b;
    let stitcherFfmpeg;
    let preStitcher;
    let encodedFrames = 0;
    let renderedFrames = 0;
    let preEncodedFileLocation;
    let renderedDoneIn = null;
    let encodedDoneIn = null;
    const renderStart = Date.now();
    if (parallelEncoding) {
        if (typeof crf !== 'number') {
            throw new TypeError('CRF is unexpectedly not a number');
        }
        preEncodedFileLocation = path_1.default.join(outputDir, 'pre-encode.' + fileExtension);
        preStitcher = await (0, stitcher_1.spawnFfmpeg)({
            dir: outputDir,
            width: config.width,
            height: config.height,
            fps: config.fps,
            outputLocation: preEncodedFileLocation,
            force: true,
            imageFormat,
            pixelFormat,
            codec,
            proResProfile,
            crf,
            parallelism,
            onProgress: (frame) => {
                encodedFrames = frame;
                onProgress({
                    encodedFrames,
                    renderedFrames,
                    encodedDoneIn: null,
                    renderedDoneIn: null,
                    stitchStage: 'encoding',
                });
            },
            verbose: remotion_1.Internals.Logging.isEqualOrBelowLogLevel('verbose'),
            parallelEncoding,
            webpackBundle: bundled,
            ffmpegExecutable,
            assetsInfo: { assets: [] },
        });
        stitcherFfmpeg = preStitcher.task;
    }
    const renderer = (0, render_1.renderFrames)({
        config,
        onFrameUpdate: (frame) => {
            renderedFrames = frame;
            onProgress({
                encodedFrames,
                renderedFrames,
                encodedDoneIn,
                renderedDoneIn,
                stitchStage: 'encoding',
            });
        },
        parallelism,
        parallelEncoding,
        outputDir,
        onStart: () => {
            renderedFrames = 0;
            onProgress({
                encodedDoneIn,
                encodedFrames,
                renderedDoneIn,
                renderedFrames,
                // TODO: Keep a state
                stitchStage: 'encoding',
            });
        },
        inputProps,
        envVariables,
        imageFormat,
        quality,
        browser,
        frameRange: frameRange !== null && frameRange !== void 0 ? frameRange : null,
        puppeteerInstance: openedBrowser,
        writeFrame: async (buffer) => {
            var _a;
            (_a = stitcherFfmpeg === null || stitcherFfmpeg === void 0 ? void 0 : stitcherFfmpeg.stdin) === null || _a === void 0 ? void 0 : _a.write(buffer);
        },
        serveUrl,
    });
    const { assetsInfo } = await renderer;
    if (stitcherFfmpeg) {
        (_a = stitcherFfmpeg === null || stitcherFfmpeg === void 0 ? void 0 : stitcherFfmpeg.stdin) === null || _a === void 0 ? void 0 : _a.end();
        await stitcherFfmpeg;
        (_b = preStitcher === null || preStitcher === void 0 ? void 0 : preStitcher.cleanup) === null || _b === void 0 ? void 0 : _b.call(preStitcher);
    }
    const closeBrowserPromise = openedBrowser.close();
    renderedDoneIn = Date.now() - renderStart;
    onProgress({
        encodedFrames,
        renderedFrames,
        renderedDoneIn,
        encodedDoneIn,
        stitchStage: 'encoding',
    });
    if (process.env.DEBUG) {
        remotion_1.Internals.perf.logPerf();
    }
    if (shouldOutputImageSequence) {
        return;
    }
    if (typeof crf !== 'number') {
        throw new TypeError('CRF is unexpectedly not a number');
    }
    const dirName = path_1.default.dirname(absoluteOutputFile);
    if (!fs_1.default.existsSync(dirName)) {
        fs_1.default.mkdirSync(dirName, {
            recursive: true,
        });
    }
    const stitchStart = Date.now();
    await (0, stitcher_1.stitchFramesToVideo)({
        dir: outputDir,
        width: config.width,
        height: config.height,
        fps: config.fps,
        outputLocation: absoluteOutputFile,
        preEncodedFileLocation,
        force: overwrite,
        imageFormat,
        pixelFormat,
        codec,
        proResProfile,
        crf,
        assetsInfo,
        parallelism,
        ffmpegExecutable,
        onProgress: (frame) => {
            onProgress({
                encodedFrames: frame,
                renderedFrames,
                renderedDoneIn,
                encodedDoneIn,
                stitchStage: 'muxing',
            });
        },
        // TODO: Optimization, Now can download before!
        onDownload,
        webpackBundle: bundled,
        verbose: remotion_1.Internals.Logging.isEqualOrBelowLogLevel('verbose'),
    });
    encodedDoneIn = Date.now() - stitchStart;
    onProgress({
        encodedDoneIn,
        encodedFrames,
        renderedDoneIn,
        renderedFrames,
        stitchStage: 'muxing',
    });
    await closeBrowserPromise;
};
exports.renderVideo = renderVideo;
