// @remotion/renderer/dist/stitcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.stitchFramesToVideo = exports.spawnFfmpeg = void 0;
const execa_1 = __importDefault(require("execa"));
const remotion_1 = require("remotion");
const assets_to_ffmpeg_inputs_1 = require("./assets-to-ffmpeg-inputs");
const calculate_asset_positions_1 = require("./assets/calculate-asset-positions");
const convert_assets_to_file_urls_1 = require("./assets/convert-assets-to-file-urls");
const download_and_map_assets_to_file_1 = require("./assets/download-and-map-assets-to-file");
const get_asset_audio_details_1 = require("./assets/get-asset-audio-details");
const calculate_ffmpeg_filters_1 = require("./calculate-ffmpeg-filters");
const create_ffmpeg_complex_filter_1 = require("./create-ffmpeg-complex-filter");
const get_audio_codec_name_1 = require("./get-audio-codec-name");
const get_codec_name_1 = require("./get-codec-name");
const get_prores_profile_name_1 = require("./get-prores-profile-name");
const parse_ffmpeg_progress_1 = require("./parse-ffmpeg-progress");
const sample_rate_1 = require("./sample-rate");
const validate_even_dimensions_with_codec_1 = require("./validate-even-dimensions-with-codec");
const validate_ffmpeg_1 = require("./validate-ffmpeg");
const getAssetsData = async (options) => {
    var _a, _b;
    const codec = (_a = options.codec) !== null && _a !== void 0 ? _a : remotion_1.Internals.DEFAULT_CODEC;
    const encoderName = (0, get_codec_name_1.getCodecName)(codec);
    const isAudioOnly = encoderName === null;
    const fileUrlAssets = await (options.assetsInfo
        ? (0, convert_assets_to_file_urls_1.convertAssetsToFileUrls)({
            assets: options.assetsInfo.assets,
            downloadDir: options.assetsInfo.downloadDir,
            onDownload: (_b = options.onDownload) !== null && _b !== void 0 ? _b : (() => () => undefined),
        })
        : null);
    (0, download_and_map_assets_to_file_1.markAllAssetsAsDownloaded)();
    const assetPositions = fileUrlAssets === null ? [] : (0, calculate_asset_positions_1.calculateAssetPositions)(fileUrlAssets);
    (0, download_and_map_assets_to_file_1.markAllAssetsAsDownloaded)();
    const assetAudioDetails = await (0, get_asset_audio_details_1.getAssetAudioDetails)({
        assetPaths: assetPositions.map((a) => a.src),
    });
    const filters = (0, calculate_ffmpeg_filters_1.calculateFfmpegFilters)({
        assetAudioDetails,
        assetPositions,
        fps: options.fps,
        videoTrackCount: isAudioOnly ? 0 : 1,
    });
    if (options.verbose) {
        console.log('asset positions', assetPositions);
    }
    if (options.verbose) {
        console.log('filters', filters);
    }
    const { complexFilterFlag, cleanup } = await (0, create_ffmpeg_complex_filter_1.createFfmpegComplexFilter)(filters);
    return {
        complexFilterFlag,
        cleanup,
        assetPositions,
    };
};
const spawnFfmpeg = async (options) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    remotion_1.Internals.validateDimension(options.height, 'height', 'passed to `stitchFramesToVideo()`');
    remotion_1.Internals.validateDimension(options.width, 'width', 'passed to `stitchFramesToVideo()`');
    remotion_1.Internals.validateFps(options.fps, 'passed to `stitchFramesToVideo()`');
    const codec = (_a = options.codec) !== null && _a !== void 0 ? _a : remotion_1.Internals.DEFAULT_CODEC;
    (0, validate_even_dimensions_with_codec_1.validateEvenDimensionsWithCodec)({
        width: options.width,
        height: options.height,
        codec,
        scale: 1,
    });
    const crf = (_b = options.crf) !== null && _b !== void 0 ? _b : remotion_1.Internals.getDefaultCrfForCodec(codec);
    const pixelFormat = (_c = options.pixelFormat) !== null && _c !== void 0 ? _c : remotion_1.Internals.DEFAULT_PIXEL_FORMAT;
    await (0, validate_ffmpeg_1.validateFfmpeg)((_d = options.ffmpegExecutable) !== null && _d !== void 0 ? _d : null);
    const encoderName = (0, get_codec_name_1.getCodecName)(codec);
    const audioCodecName = (0, get_audio_codec_name_1.getAudioCodecName)(codec);
    const proResProfileName = (0, get_prores_profile_name_1.getProResProfileName)(codec, options.proResProfile);
    const isAudioOnly = encoderName === null;
    const supportsCrf = encoderName && codec !== 'prores';
    if (options.verbose) {
        console.log('[verbose] ffmpeg', (_e = options.ffmpegExecutable) !== null && _e !== void 0 ? _e : 'ffmpeg in PATH');
        console.log('[verbose] encoder', encoderName);
        console.log('[verbose] audioCodec', audioCodecName);
        console.log('[verbose] pixelFormat', pixelFormat);
        if (supportsCrf) {
            console.log('[verbose] crf', crf);
        }
        console.log('[verbose] codec', codec);
        console.log('[verbose] isAudioOnly', isAudioOnly);
        console.log('[verbose] proResProfileName', proResProfileName);
    }
    remotion_1.Internals.validateSelectedCrfAndCodecCombination(crf, codec);
    remotion_1.Internals.validateSelectedPixelFormatAndCodecCombination(pixelFormat, codec);
    const { complexFilterFlag = undefined, cleanup = undefined, assetPositions = undefined, } = ((_f = options.internalOptions) === null || _f === void 0 ? void 0 : _f.parallelEncoding)
        ? {}
        : await getAssetsData(options);
    const ffmpegArgs = [
        ['-r', String(options.fps)],
        ...(((_g = options.internalOptions) === null || _g === void 0 ? void 0 : _g.preEncodedFileLocation)
            ? [['-i', (_h = options.internalOptions) === null || _h === void 0 ? void 0 : _h.preEncodedFileLocation]]
            : isAudioOnly
                ? []
                : [
                    [
                        '-f',
                        ((_j = options.internalOptions) === null || _j === void 0 ? void 0 : _j.parallelEncoding)
                            ? ['image2pipe']
                            : 'image2',
                    ],
                    ['-s', `${options.width}x${options.height}`],
                    options.assetsInfo
                        ? ['-start_number', String(options.assetsInfo.firstFrameIndex)]
                        : null,
                    options.assetsInfo
                        ? ['-i', options.assetsInfo.imageSequenceName]
                        : null,
                    // If scale is very small (like 0.1), FFMPEG cannot figure out the image
                    // format on it's own and we need to hint the format
                    ((_k = options.internalOptions) === null || _k === void 0 ? void 0 : _k.parallelEncoding)
                        ? [
                            '-vcodec',
                            options.internalOptions.imageFormat === 'jpeg'
                                ? 'mjpeg'
                                : 'png',
                        ]
                        : null,
                    ((_l = options.internalOptions) === null || _l === void 0 ? void 0 : _l.parallelEncoding) ? ['-i', '-'] : null,
                ]),
        ...(options.assetsInfo && assetPositions
            ? (0, assets_to_ffmpeg_inputs_1.assetsToFfmpegInputs)({
                assets: assetPositions.map((a) => a.src),
                isAudioOnly,
                fps: options.fps,
                frameCount: options.assetsInfo.assets.length,
            })
            : []),
        encoderName
            ? // -c:v is the same as -vcodec as -codec:video
                // and specified the video codec.
                ['-c:v', encoderName]
            : // If only exporting audio, we drop the video explicitly
                ['-vn'],
        ...(((_m = options.internalOptions) === null || _m === void 0 ? void 0 : _m.preEncodedFileLocation)
            ? []
            : [
                proResProfileName ? ['-profile:v', proResProfileName] : null,
                supportsCrf ? ['-crf', String(crf)] : null,
                isAudioOnly ? null : ['-pix_fmt', pixelFormat],
                // Without explicitly disabling auto-alt-ref,
                // transparent WebM generation doesn't work
                pixelFormat === 'yuva420p' ? ['-auto-alt-ref', '0'] : null,
                isAudioOnly ? null : ['-b:v', '1M'],
            ]),
        '-ar',
        String(sample_rate_1.DEFAULT_SAMPLE_RATE),
        // Stereo sound, even force mono to be stereo
        // Otherwise mixing mono + stereo ends up speeding up the audio
        '-ac',
        '2',
        audioCodecName ? ['-c:a', audioCodecName] : null,
        complexFilterFlag,
        // Ignore audio from image sequence
        isAudioOnly ? null : ['-map', '0:v'],
        // Ignore metadata that may come from remote media
        isAudioOnly ? null : ['-map_metadata', '-1'],
        options.force ? '-y' : null,
        options.outputLocation,
    ];
    if (options.verbose) {
        console.log('Generated FFMPEG command:');
        console.log(ffmpegArgs);
    }
    const ffmpegString = ffmpegArgs.flat(2).filter(Boolean);
    const task = (0, execa_1.default)((_o = options.ffmpegExecutable) !== null && _o !== void 0 ? _o : 'ffmpeg', ffmpegString, {
        cwd: ((_p = options.internalOptions) === null || _p === void 0 ? void 0 : _p.parallelEncoding) ? undefined : options.dir,
    });
    let ffmpegOutput = '';
    (_q = task.stderr) === null || _q === void 0 ? void 0 : _q.on('data', (data) => {
        const str = data.toString();
        ffmpegOutput += str;
        if (options.onProgress) {
            const parsed = (0, parse_ffmpeg_progress_1.parseFfmpegProgress)(str);
            if (parsed !== undefined) {
                options.onProgress(parsed);
            }
        }
    });
    return { task, cleanup, getLogs: () => ffmpegOutput };
};
exports.spawnFfmpeg = spawnFfmpeg;
const stitchFramesToVideo = async (options) => {
    const { task, cleanup, getLogs } = await (0, exports.spawnFfmpeg)(options);
    try {
        await task;
    }
    catch (err) {
        throw new Error(getLogs());
    }
    finally {
        cleanup === null || cleanup === void 0 ? void 0 : cleanup();
    }
};
exports.stitchFramesToVideo = stitchFramesToVideo;
