// @remotion/cli/dist/editor/components/NewComposition/CompositionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompositionType = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../../helpers/colors");
const film_1 = require("../../icons/film");
const still_1 = require("../../icons/still");
const z_index_1 = require("../../state/z-index");
const layout_1 = require("../layout");
const panel = {
    flex: 1,
    borderRadius: 8,
    padding: 16,
    cursor: 'pointer',
    fontSize: 14,
    flexDirection: 'row',
    display: 'flex',
    alignItems: 'center',
};
const Panel = ({ selected, type, onSelected }) => {
    const ref = (0, react_1.useRef)(null);
    const { tabIndex } = (0, z_index_1.useZIndex)();
    const [hovered, setHovered] = (0, react_1.useState)(false);
    (0, react_1.useEffect)(() => {
        const { current } = ref;
        if (!current) {
            return;
        }
        const onMouseEnter = () => setHovered(true);
        const onMouseLeave = () => setHovered(false);
        current.addEventListener('mouseenter', onMouseEnter);
        current.addEventListener('mouseleave', onMouseLeave);
    }, []);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...panel,
            backgroundColor: (0, colors_1.getBackgroundFromHoverState)({
                hovered,
                selected,
            }),
            color: 'white',
            appearance: 'none',
            border: 'none',
            textAlign: 'left',
        };
    }, [hovered, selected]);
    const onClick = (0, react_1.useCallback)(() => {
        onSelected(type);
    }, [onSelected, type]);
    return ((0, jsx_runtime_1.jsxs)("button", Object.assign({ ref: ref, tabIndex: tabIndex, type: "button", style: style, onClick: onClick }, { children: [type === 'composition' ? ((0, jsx_runtime_1.jsx)(film_1.FilmIcon, { style: { height: 18, width: 18 } }, void 0)) : ((0, jsx_runtime_1.jsx)(still_1.StillIcon, { style: { height: 18, width: 18 } }, void 0)), (0, jsx_runtime_1.jsx)(layout_1.Spacing, { x: 1 }, void 0), (0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)("strong", { children: type === 'composition' ? '<Composition>' : '<Still>' }, void 0), (0, jsx_runtime_1.jsx)("div", { children: type === 'composition' ? 'For videos' : 'For still images' }, void 0)] }, void 0)] }), void 0));
};
const CompositionType = ({ onSelected, type }) => {
    return ((0, jsx_runtime_1.jsxs)(layout_1.Row, { children: [(0, jsx_runtime_1.jsx)(Panel, { onSelected: onSelected, type: "composition", selected: type === 'composition' }, void 0), (0, jsx_runtime_1.jsx)(Panel, { onSelected: onSelected, type: "still", selected: type === 'still' }, void 0)] }, void 0));
};
exports.CompositionType = CompositionType;
