// @remotion/bundler/dist/error-overlay/remotion-overlay/ErrorTitle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorTitle = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_overlay_1 = require("../react-overlay");
const DismissButton_1 = require("./DismissButton");
const Symbolicating_1 = require("./Symbolicating");
const title = {
    marginBottom: 8,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
};
const left = {
    flex: 1,
    paddingRight: 14,
    lineHeight: 1.5,
    whiteSpace: 'pre',
    fontSize: '1.5em',
    fontWeight: 'bold',
    overflowX: 'auto',
};
const errName = {
    fontSize: '0.8em',
    background: 'linear-gradient(90deg,#4290f5,#42e9f5)',
    WebkitBackgroundClip: 'text',
    WebkitTextFillColor: 'transparent',
    display: 'inline-block',
};
const row = {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
};
const spacer = {
    width: 5,
};
const ErrorTitle = ({ name, message, symbolicating }) => {
    return ((0, jsx_runtime_1.jsxs)("div", { style: title, children: [(0, jsx_runtime_1.jsxs)("div", { style: left, children: [(0, jsx_runtime_1.jsx)("span", { style: errName, children: name }, void 0), (0, jsx_runtime_1.jsx)("br", {}, void 0), (0, jsx_runtime_1.jsxs)("div", { style: row, children: [symbolicating ? ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Symbolicating_1.Symbolicating, {}, void 0), (0, jsx_runtime_1.jsx)("div", { style: spacer }, void 0)] }, void 0)) : null, (0, jsx_runtime_1.jsx)("div", { children: message }, void 0)] }, void 0)] }, void 0), (0, react_overlay_1.didUnmountReactApp)() ? null : (0, jsx_runtime_1.jsx)(DismissButton_1.DismissButton, {}, void 0)] }, void 0));
};
exports.ErrorTitle = ErrorTitle;
