// @remotion/cli/dist/editor/components/Errors/ErrorDisplay.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorDisplay = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const container = {
    width: '100%',
    maxWidth: 1000,
    marginLeft: 'auto',
    marginRight: 'auto',
    fontFamily: 'sans-serif',
    paddingTop: '5vh',
};
const title = {
    fontSize: '1.5em',
    fontWeight: 'bold',
};
const stack = {
    backgroundColor: 'black',
    marginTop: 17,
};
const stackLine = {
    paddingTop: 8,
    paddingBottom: 8,
    paddingLeft: 14,
    paddingRight: 14,
};
const ErrorDisplay = ({ display }) => {
    return ((0, jsx_runtime_1.jsxs)("div", Object.assign({ style: container }, { children: [(0, jsx_runtime_1.jsxs)("div", Object.assign({ style: title }, { children: [display.error.name, ": ", display.error.message] }), void 0), (0, jsx_runtime_1.jsx)("div", Object.assign({ style: stack }, { children: display.stackFrames.map((s, i) => {
                    var _a;
                    return (
                    // eslint-disable-next-line react/no-array-index-key
                    (0, jsx_runtime_1.jsxs)("div", Object.assign({ style: stackLine }, { children: [(0, jsx_runtime_1.jsx)("div", { children: s.functionName }, void 0), (0, jsx_runtime_1.jsxs)("div", { children: [s._originalFileName, ":", s.columnNumber] }, void 0), (0, jsx_runtime_1.jsx)("div", { children: (_a = s._originalScriptCode) === null || _a === void 0 ? void 0 : _a.map((s, j) => {
                                    return ((0, jsx_runtime_1.jsx)("div", Object.assign({ style: {
                                            fontFamily: 'monospace',
                                            whiteSpace: 'pre',
                                            tabSize: 2,
                                            color: s.highlight ? 'red' : 'inherit',
                                        } }, { children: s.content }), j));
                                }) }, void 0)] }), i));
                }) }), void 0)] }), void 0));
};
exports.ErrorDisplay = ErrorDisplay;
